tls/s2n_protocol_preferences.c

```diff
@@ -17,39 +17,89 @@
+
+    uint32_t extension_len = s2n_stuffer_data_available(protocol_stuffer) + /* len prefix */ 1 + protocol_len;
+    ENSURE(extension_len <= UINT16_MAX, S2N_ERR_INVALID_APPLICATION_PROTOCOL);
+
+    GUARD_AS_RESULT(s2n_stuffer_write_uint8(protocol_stuffer, protocol_len));
+    GUARD_AS_RESULT(s2n_stuffer_write_bytes(protocol_stuffer, protocol, protocol_len));
 
-    GUARD(s2n_free(application_protocols));
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_protocol_preferences_set(struct s2n_blob *application_protocols, const char *const *protocols, int protocol_count)
+{
+    ENSURE_MUT(application_protocols);
 
+    /* NULL value indicates no preference so free the previous blob */
     if (protocols == NULL || protocol_count == 0) {
-        /* NULL value indicates no preference, so nothing to do */
-        return 0;
+        GUARD_AS_RESULT(s2n_free(application_protocols));
+
+        return S2N_RESULT_OK;
     }
 
-    GUARD(s2n_stuffer_growable_alloc(&protocol_stuffer, 256));
-    for (int i = 0; i < protocol_count; i++) {
-        size_t length = strlen(protocols[i]);
-        uint8_t protocol[255];
+    DEFER_CLEANUP(struct s2n_stuffer protocol_stuffer = { 0 }, s2n_stuffer_free);
 
-        S2N_ERROR_IF(length > 255 || (s2n_stuffer_data_available(&protocol_stuffer) + length + 1) > 65535, S2N_ERR_APPLICATION_PROTOCOL_TOO_LONG);
-        memcpy_check(protocol, protocols[i], length);
-        GUARD(s2n_stuffer_write_uint8(&protocol_stuffer, length));
-        GUARD(s2n_stuffer_write_bytes(&protocol_stuffer, protocol, length));
+    GUARD_AS_RESULT(s2n_stuffer_growable_alloc(&protocol_stuffer, 256));
+    for (size_t i = 0; i < protocol_count; i++) {
+        const uint8_t * protocol = (const uint8_t *)protocols[i];
+        size_t length = strlen(protocols[i]);
+        GUARD_RESULT(s2n_protocol_preferences_write(&protocol_stuffer, protocol, length));
     }
 
-    GUARD(s2n_stuffer_extract_blob(&protocol_stuffer, application_protocols));
-    GUARD(s2n_stuffer_free(&protocol_stuffer));
-    return 0;
+    GUARD_AS_RESULT(s2n_stuffer_extract_blob(&protocol_stuffer, application_protocols));
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_protocol_preferences_append(struct s2n_blob *application_protocols, const uint8_t *protocol, size_t protocol_len)
+{
+    ENSURE_MUT(application_protocols);
+    ENSURE_REF(protocol);
+
+    struct s2n_stuffer protocol_stuffer = {0};
+    GUARD_AS_RESULT(s2n_stuffer_init(&protocol_stuffer, application_protocols));
+    protocol_stuffer.alloced = true;
+    protocol_stuffer.growable = true;
+    GUARD_AS_RESULT(s2n_stuffer_skip_write(&protocol_stuffer, application_protocols->size));
+
+    GUARD_RESULT(s2n_protocol_preferences_write(&protocol_stuffer, protocol, protocol_len));
+
+    /* ensure the application_protocols gets updated if we reallocate */
+    *application_protocols = protocol_stuffer.blob;
+    application_protocols->size = protocol_stuffer.write_cursor;
+
```

Comment:
Wouldn't this cause a memory leak? On a resize, a stuffer could (and [should](https://github.com/awslabs/s2n/blob/main/stuffer/s2n_stuffer.c#L370)) allocate more memory that it needs. You could be telling a blob to forget about some of its allocated memory

Reply:
Nope since stuffer takes a shallow copy of a blob. But since this pattern isn't used anywhere I'll just use the blob api directly